### PR TITLE
Cosmos: Add embedded collection support

### LINKED ISSUE #16620: Nav Expansion: Owned collections are expanded
Owned navigation cannot be expanded

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Flatten out shaper expression visitors
Don't expand owned collection

Fixes #16620
Part of #12086

### CODE REVIEW

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs [hunk 1]
```diff
@@ -196,12 +196,13 @@ private class EntityMaterializerInjectingExpressionVisitor : ExpressionVisitor
 
             protected override Expression VisitExtension(Expression extensionExpression)
             {
-                if (extensionExpression is EntityShaperExpression entityShaperExpression)
+                switch (extensionExpression)
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: conditional return statement?

File: src/EFCore.Cosmos/Query/Pipeline/ArrayProjectionExpression.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+
+namespace Microsoft.EntityFrameworkCore.Cosmos.Query.Pipeline
+{
+    public class ArrayProjectionExpression : Expression, IAccessExpression
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/ArrayProjectionExpression.cs]: IPrintable

Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Pipeline/ArrayProjectionExpression.cs]: Not sure what to print other than the contained `EntityProjectionExpression`...

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: I'll just revert to `if`

File: src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs [hunk 3]
```diff
@@ -176,170 +316,134 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
+                        var counterParameter = Expression.Parameter(typeof(int), "counterParameter");
 
-                            _jObjectParameter = parentJObject;
+                        _projectionBindings[arrayProjection.EntityExpression.AccessExpression] = jObjectParameter;
 
-                            expressions.Add(Expression.Condition(
-                                Expression.Equal(jObjectVariable, Expression.Constant(null, jObjectVariable.Type)),
-                                Expression.Constant(null, shaperExpression.Type),
-                                materializerExpression));
+                        var innerShaper = base.Visit(collectionShaperExpression.InnerShaper);
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs]: Is base necessary here?

File: src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs [hunk 4]
```diff
@@ -176,170 +316,134 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
+                                navigationExpression = Expression.Assign(
+                                    Expression.MakeMemberAccess(
+                                        convertedInstance,
+                                        navigationMemberInfo),
+                                    navigationExpression);
+                            }
+                        } else if (navigation.IsCollection())
+                        {
+                            navigationExpression = Expression.Call(
+                                _loadMethodInfo.MakeGenericMethod(navigation.GetTargetType().ClrType),
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs]: Why do we have to call Load to iterate the collection? Shouldn't materialization of each entity in collection add it to statemanager to fix it up?

File: src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs [hunk 5]
```diff
@@ -176,170 +316,134 @@ protected override Expression VisitExtension(Expression extensionExpression)
...
-                var instanceVariable = expressions[expressions.Count - 1];
-                expressions.RemoveAt(expressions.Count - 1);
+                        var navigationMemberInfo = navigation.GetMemberInfo(forConstruction: true, forSet: true);
+                        var convertedInstance = navigationMemberInfo.DeclaringType.IsAssignableFrom(instanceVariable.Type)
+                            ? instanceVariable
+                            : Expression.Convert(instanceVariable, navigationMemberInfo.DeclaringType);
 
-                var concreteEntityTypeVariableName = "entityType" + _currentEntityIndex;
-                var concreteEntityTypeVariable = blockExpression.Variables.Single(v => v.Name == concreteEntityTypeVariableName);
+                        if (!_trackQueryResults)
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs]: Also you need to set IsLoaded flags for navigations

File: src/EFCore.Cosmos/Query/Pipeline/CosmosSqlTranslatingExpressionVisitor.cs [hunk 6]
```diff
@@ -339,6 +314,9 @@ protected override Expression VisitExtension(Expression extensionExpression)
                 case NullConditionalExpression nullConditionalExpression:
                     return Visit(nullConditionalExpression.AccessOperation);
 
+                case IncludeExpression includeExpression:
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosSqlTranslatingExpressionVisitor.cs]: It should not come here. Server translation of IncludeExpression should be null

File: src/EFCore.Cosmos/Query/Pipeline/ArrayProjectionExpression.cs (hunk 2 above)
Comment (smitpatel) [reply src/EFCore.Cosmos/Query/Pipeline/ArrayProjectionExpression.cs]: Well that also helps, when debugging. If you don't add it boss will do eventually :trollface:

File: src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs (hunk 4 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Pipeline/CosmosShapedQueryCompilingExpressionVisitor.cs]: They will only be materialized if the collection materializer is enumerated.

Conceptually `navigationExpression` is `jArray.Select(o => Materialize(o))`

**Comment by AndriySvyryd** [comment]
@smitpatel Updated and rebased

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs [hunk 7]
```diff
@@ -246,14 +239,14 @@ private Expression ProcessEntityShaper(EntityShaperExpression entityShaperExpres
 
                 if (_trackQueryResults)
                 {
-                    var entry = Expression.Variable(typeof(InternalEntityEntry), "entry" + _currentEntityIndex);
+                    var entryVariable = Expression.Variable(typeof(InternalEntityEntry), "entry" + _currentEntityIndex);
                     var hasNullKey = Expression.Variable(typeof(bool), "hasNullKey" + _currentEntityIndex);
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: hasNullKeyVariable? just to be consistent.

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs [hunk 8]
```diff
@@ -275,16 +268,17 @@ private Expression ProcessEntityShaper(EntityShaperExpression entityShaperExpres
...
                             Expression.Block(
-                                Expression.Assign(instanceVariable, Expression.Convert(
-                                    Expression.MakeMemberAccess(entry, _entityMemberInfo),
-                                    entityType.ClrType)),
                                 Expression.Assign(
                                     concreteEntityTypeVariable,
-                                    Expression.MakeMemberAccess(entry, _entityTypeMemberInfo))),
-                            MaterializeEntity(entityType, materializationContextVariable, concreteEntityTypeVariable, instanceVariable))));
+                                    Expression.MakeMemberAccess(entryVariable, _entityTypeMemberInfo)),
+                                Expression.Assign(instanceVariable, Expression.Convert(
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: Any significance of different ordering now?

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs [hunk 9]
```diff
@@ -275,16 +268,17 @@ private Expression ProcessEntityShaper(EntityShaperExpression entityShaperExpres
...
+                                    Expression.MakeMemberAccess(entryVariable, _entityMemberInfo),
+                                    entityType.ClrType))),
+                            MaterializeEntity(
+                                entityType, materializationContextVariable, concreteEntityTypeVariable, instanceVariable, entryVariable))));
```
Comment (smitpatel) [inline src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: How are we using entry variable

File: src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs [hunk 10]
```diff
@@ -59,68 +61,63 @@ public override Expression Visit(Expression expression)
...
-                    {
+                    case MaterializeCollectionNavigationExpression materializeCollectionNavigationExpression:
                         return base.Visit(expression);
-                    }
-                    else
-                    {
-                        return new ProjectionBindingExpression(
-                            _selectExpression, _selectExpression.AddToProjection(translation), expression.Type);
-                    }
+                    //return _selectExpression.AddCollectionProjection(
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs]: Should remove commented code

File: src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs [hunk 11]
```diff
@@ -132,39 +129,123 @@ private static T GetParameterValue<T>(QueryContext queryContext, string paramete
...
 
-        protected override Expression VisitExtension(Expression extensionExpression)
+        protected override Expression VisitMember(MemberExpression memberExpression)
         {
-            if (extensionExpression is EntityShaperExpression entityShaperExpression)
+            if (!_clientEval)
             {
-                var projectionBindingExpression = (ProjectionBindingExpression)entityShaperExpression.ValueBufferExpression;
-                VerifySelectExpression(projectionBindingExpression);
+                return base.VisitMember(memberExpression);
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs]: Should this call base or return null?
if we are not in client eval phase, then shouldn't the translation of any component (direct expression or anonymous/nominal type's components) should go through SqlTranslator?

File: src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs [hunk 12]
```diff
@@ -132,39 +129,123 @@ private static T GetParameterValue<T>(QueryContext queryContext, string paramete
...
+            }
 
-                if (_clientEval)
-                {
-                    var entityProjection = (EntityProjectionExpression)_selectExpression.GetMappedProjection(
-                        projectionBindingExpression.ProjectionMember);
+            var newExpression = Visit(memberExpression.Expression);
```
Comment (smitpatel) [inline src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs]: innerExpression? visitedExpression?

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs (hunk 8 above)
Comment (AndriySvyryd) [reply src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: Consistency. The other blocks end with the instance.

File: src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs (hunk 9 above)
Comment (AndriySvyryd) [reply src/EFCore/Query/Pipeline/ShapedQueryExpressionVisitor.cs]: To set `IsLoaded` later

File: src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs (hunk 10 above)
Comment (AndriySvyryd) [reply src/EFCore.Cosmos/Query/Pipeline/CosmosProjectionBindingExpressionVisitor.cs]: But we'll need it for non-cross-partition joins

**Pull request merged**